File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/JavaFxWorker.java
package com.cedarsoft.commons.javafx;

import java.util.concurrent.ExecutionException;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import com.cedarsoft.annotations.NonUiThread;
import com.cedarsoft.annotations.UiThread;

import javafx.application.Platform;
import javafx.concurrent.Task;

/**
 * Similar to a simple SwingWorker.
 * Must not return "null". Instead use Optional
 *
 * @author Johannes Schneider (<a href="mailto:js@cedarsoft.com">js@cedarsoft.com</a>)
 */
public abstract class JavaFxWorker<T> extends Task<T> {
  @Nullable
  @Override
  protected final T call() throws Exception {
    return doInBackground();
  }

  @Nullable
  protected abstract T doInBackground() throws Exception;

  @UiThread
  @Override
  protected final void succeeded() {
    super.succeeded();
  }

  @UiThread
  @Override
  protected final void running() {
    super.running();
  }

  @UiThread
  @Override
  protected final void cancelled() {
    super.cancelled();
  }

  @UiThread
  @Override
  protected final void failed() {
    super.failed();
  }

  @NonUiThread
  @Override
  protected final void done() {
    super.done();

    Platform.runLater(() -> {
      cleanup();
      try {
        success(get());
      }
      catch (InterruptedException e) {
        failure(e);
      }
      catch (ExecutionException e) {
        failure(e.getCause());
      }
      catch (Exception e) {
        failure(e);
      }
    });
  }

  /**
   * Is called before #done(T);
   * This method is called even if {@link #doInBackground()} has thrown an exception
   */
  @UiThread
  protected void cleanup() {
  }

  /**
   * Is called the value that is returned by {@link #doInBackground()}.
   * Is only called if no exception has been thrown in {@link #doInBackground()}.
   */
  @UiThread
  protected void success(@Nullable T t) {
  }

  @UiThread
  protected void failure(@Nonnull Throwable e) {
    throw new RuntimeException(e);
  }

  /**
   * Schedules for execution
   *
   * @return the thread this worker runs its background stuff on
   */
  @Nonnull
  public Thread execute() {
    Thread thread = new Thread(this, "JavaFXWorker-Thread for <" + getClass().getName() + ">");
    thread.start();
    return thread;
  }
}

File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/JavaFxWorker.kt
package com.cedarsoft.commons.javafx

import com.cedarsoft.annotations.NonUiThread
import com.cedarsoft.annotations.UiThread
import com.cedarsoft.unit.si.t
import javafx.application.Platform
import javafx.concurrent.Task
import java.util.concurrent.ExecutionException
import javax.annotation.Nonnull

/**
 * Similar to a simple SwingWorker.
 * Must not return "null". Instead, use [java.util.Optional]
 *
 */
abstract class JavaFxWorker<T> : Task<T>() {
  @Throws(Exception::class)
  override fun call(): T? {
    return doInBackground()
  }

  @Throws(Exception::class)
  protected abstract fun doInBackground(): T

  @UiThread
  final override fun succeeded() {
    super.succeeded()
  }

  @UiThread
  final override fun running() {
    super.running()
  }

  @UiThread
  final override fun cancelled() {
    super.cancelled()
  }

  @UiThread
  final override fun failed() {
    super.failed()
  }

  @NonUiThread
  override fun done() {
    super.done()
    Platform.runLater {
      cleanup()
      try {
        success(get())
      } catch (e: InterruptedException) {
        failure(e)
      } catch (e: ExecutionException) {
        failure(e.cause ?: e)
      } catch (e: Exception) {
        failure(e)
      }
    }
  }

  /**
   * Is called before #done(T);
   * This method is called even if [.doInBackground] has thrown an exception
   */
  @UiThread
  protected open fun cleanup() {
  }

  /**
   * Is called the value that is returned by [.doInBackground].
   * Is only called if no exception has been thrown in [.doInBackground].
   */
  @UiThread
  protected open fun success(t: T) {
  }

  @UiThread
  protected open fun failure(e: Throwable) {
    throw e
  }

  /**
   * Schedules for execution
   *
   * @return the thread this worker runs its background stuff on
   */
  @Nonnull
  fun execute(): Thread {
    val thread = Thread(this, "JavaFXWorker-Thread for <" + javaClass.name + ">")
    thread.start()
    return thread
  }

  companion object {
    /**
     * Creates a very simple fx worker
     */
    fun <T> create(backgroundAction: () -> T): JavaFxWorker<T> {
      return object : JavaFxWorker<T>() {
        override fun doInBackground(): T {
          return backgroundAction()
        }
      }
    }

    fun <T> build(configuration: Builder<T>.() -> Unit): Builder<T> {
      return Builder<T>().also {
        it.configuration()
      }
    }

    operator fun <T> invoke(configuration: Builder<T>.() -> Unit) {
      build(configuration)
        .build()
        .execute()
    }
  }

  /**
   * A builder for a javafx worker
   */
  class Builder<T> {
    var backgroundAction: (() -> T)? = null

    var onCleanup: () -> Unit = {}

    var onSuccess: (t: T?) -> Unit = {}
    var onFailure: (e: Throwable) -> Unit = {}

    fun build(): JavaFxWorker<T> {
      val backgroundActionCopy = backgroundAction
      requireNotNull(backgroundActionCopy) {
        "No backgroundAction has been set!"
      }

      return object : JavaFxWorker<T>() {
        override fun doInBackground(): T {
          return backgroundActionCopy.invoke()
        }

        override fun cleanup() {
          super.cleanup()
          onCleanup()
        }

        override fun success(t: T) {
          super.success(t)
          onSuccess(t)
        }

        override fun failure(e: Throwable) {
          super.failure(e)
          onFailure(e)
        }
      }
    }
  }
}
